src/Nattish.hs

```diff
@@ -81,4 +82,30 @@ pattern Succy n <- (check -> ResSucc n)
 
 {-# COMPLETE Zeroy, Succy #-}
 
+-- [Note: WordPtr]
+--
+-- Why WordPtr and not Word? We want to be extremely certain that it does not
+-- overflow, because that would lead to a type safety problem (accessing a Succ
+-- object as a value). At first, I thought we couldn't possibly have to worry
+-- about that—after all, the Nattish never exceeds the *logarithm* of the queue
+-- size. But then I remembered that it's possible to make quite absurdly large
+-- queues by repeated concatenation.  Imagine if a Word is 32 bits, but a
+-- pointer is 64 bits. Then something like
+--
+--   stimes (2^(2^33)) (singleton () ())
+--
+-- might actually be possible to construct in a relatively reasonable amount of
+-- time and an achievable (if somewhat ridiculous) amount of memory. If this
+-- were then traversed with traverseWithKeyU or foldMapWithKeyU from right to
+-- left (using a lazy left fold, for example), we could overflow. Disaster!
+-- Fortunately, we have WordPtr, which is equivalent to uintptr_t, and
+-- therefore big enough to hold the logarithm of the size of any queue
+-- represented in memory, even with maximal sharing.
+--
+-- Now, according to a comment on
+-- https://hackage.haskell.org/package/base-4.18.0.0/docs/src/Foreign.Ptr.html#WordPtr
+-- GHC actually guarantees that a Word is the same size as a pointer, so we
+-- actually *could* just use Word, but I think WordPtr makes it *utterly clear*
+-- that we're safe here.
```

Comment:
What? We use `WordPtr` because it could be bigger than `Word`, but actually they're always the same size? What's the point then? I think using `WordPtr` only makes it more confusing, since usually noone uses a pointer as a number type.

Reply:
On GHC they're the same. If someone ports to another implementation, that might not be.

Reply:
If you're dead set against this, I can go back to `Word` and explain why it's okay on GHC. That just commits us even more heavily to GHC *specifically*, rather than to just a compiler with similar power.